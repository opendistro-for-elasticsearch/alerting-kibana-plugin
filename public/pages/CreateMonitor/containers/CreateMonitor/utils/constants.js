import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const BUCKET_COUNT = 5;

export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);

export const FORMIK_INITIAL_VALUES = {
  /* CONFIGURE MONITOR */
  name: '',
  disabled: false,
  frequency: 'interval',
  timezone: [],
  daily: 0,
  period: { interval: 1, unit: 'MINUTES' },
  weekly: { mon: false, tue: false, wed: false, thur: false, fri: false, sat: false, sun: false },
  monthly: { type: 'day', day: 1 },
  cronExpression: '0 */1 * * *',

  /* DEFINE MONITOR */
  searchType: 'graph',
  index: [],
  timeField: '',
  query: MATCH_ALL_QUERY,
  aggregationType: 'count',
  fieldName: [],
  overDocuments: 'all documents',
  groupedOverTop: 5,
  groupedOverFieldName: 'bytes',
  bucketValue: 1,
  bucketUnitOfTime: 'h', // m = minute, h = hour, d = day
  where: {
    fieldName: [],
    operator: OPERATORS_MAP.IS,
    fieldValue: '',
    fieldRangeStart: 0,
    fieldRangeEnd: 0,
  },
  detectorId: '',
};
