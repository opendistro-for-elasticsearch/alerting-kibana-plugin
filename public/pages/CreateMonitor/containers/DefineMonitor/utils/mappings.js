/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export function shouldSkip(mapping) {
  const isDisabled = mapping.enabled === false;
  const hasIndexDisabled = mapping.index === false;
  const isNestedDataType = mapping.type === 'nested';
  return isDisabled || hasIndexDisabled || isNestedDataType;
}

export function resolvePath(path, field) {
  if (path) return `${path}.${field}`;
  return field;
}

export function getFieldsFromProperties(properties, dataTypes, path) {
  Object.entries(properties).forEach(([field, value]) => {
    getTypeFromMappings(value, dataTypes, resolvePath(path, field));
  });
}

export function getTypeFromMappings(mappings, dataTypes, path = '') {
  if (shouldSkip(mappings)) return dataTypes;
  // if there are properties then type is inherently an object
  if (mappings.properties) {
    getFieldsFromProperties(mappings.properties, dataTypes, path);
    return dataTypes;
  }

  const type = mappings.type;

  if (dataTypes[type]) dataTypes[type].add(path);
  else dataTypes[type] = new Set([path]);
  return dataTypes;
}

export function getPathsPerDataType(mappings) {
  const dataTypes = {};
  Object.entries(mappings)
    .forEach(([index, { mappings: docMappings }]) => getTypeFromMappings(docMappings, dataTypes));
  return dataTypes;
}
