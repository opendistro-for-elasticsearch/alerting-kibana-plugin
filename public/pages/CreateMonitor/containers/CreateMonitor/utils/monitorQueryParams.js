import { SEARCH_TYPE } from '../../../../../utils/constants';

export const initializeFromQueryParams = queryParams => {
  return {
    searchType: queryParams.searchType || undefined,
    name: queryParams.name ? `${queryParams.name}-Monitor` : undefined,
    detectorId: queryParams.adId || undefined,
    period:
      queryParams.interval && queryParams.unit
        ? { interval: parseInt(queryParams.interval), unit: queryParams.unit }
        : undefined,
  };
};
