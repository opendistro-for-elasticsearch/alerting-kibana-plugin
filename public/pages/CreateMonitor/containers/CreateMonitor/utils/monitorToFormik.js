/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { FORMIK_INITIAL_VALUES } from './constants';
import { SEARCH_TYPE, INPUTS_DETECTOR_ID } from '../../../../../utils/constants';

// Convert Monitor JSON to Formik values used in UI forms
export default function monitorToFormik(monitor) {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
  if (!monitor) return formikValues;
  const {
    name,
    enabled,
    schedule: { cron: { expression: cronExpression = formikValues.cronExpression, timezone } = {} },
    inputs,
    ui_metadata: { schedule = {}, search = {} } = {},
  } = monitor;
  // Default searchType to query, because if there is no ui_metadata or search then it was created through API or overwritten by API
  // In that case we don't want to guess on the UI what selections a user made, so we will default to just showing the extraction query
  const { searchType = 'query', fieldName } = search;
  const isAD = searchType === SEARCH_TYPE.AD;

  return {
    /* INITIALIZE WITH DEFAULTS */
    ...formikValues,

    /* CONFIGURE MONITOR */
    name,
    disabled: !enabled,

    /* This will overwrite the fields in use by Monitor from ui_metadata */
    ...schedule,
    cronExpression,

    /* DEFINE MONITOR */
    ...search,
    searchType,
    fieldName: fieldName ? [{ label: fieldName }] : [],
    timezone: timezone ? [{ label: timezone }] : [],

    detectorId: isAD //&& _.get(inputs, '0.search.indices.0') === '.opendistro-anomaly-results*'
      ? _.get(inputs, INPUTS_DETECTOR_ID)
      : undefined,
    index: inputs[0].search.indices.map(index => ({ label: index })),
    query: JSON.stringify(inputs[0].search.query, null, 4),
  };
}
