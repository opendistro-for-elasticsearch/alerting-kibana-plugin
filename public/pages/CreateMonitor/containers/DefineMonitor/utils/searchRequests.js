/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  formikToGraphQuery,
  formikToUiGraphQuery,
  formikToIndices,
} from '../../CreateMonitor/utils/formikToMonitor';
import { SEARCH_TYPE } from '../../../../../utils/constants';

export const buildSearchRequest = (values, uiGraphQuery = true) => {
  console.log('#values in buildSearchRequest: ', values);
  switch (values.searchType) {
    case SEARCH_TYPE.QUERY:
      return buildQuerySearchRequest(values);
    case SEARCH_TYPE.GRAPH:
      return buildGraphSearchRequest(values, uiGraphQuery);
    case SEARCH_TYPE.HTTP:
      return buildHttpRequest(values);
  }
};

console.log('#entered search request');

function buildQuerySearchRequest(values) {
  const indices = formikToIndices(values);
  const query = JSON.parse(values.query);
  return { query, indices };
}

function buildGraphSearchRequest(values, uiGraphQuery) {
  const query = uiGraphQuery ? formikToUiGraphQuery(values) : formikToGraphQuery(values);
  const indices = formikToIndices(values);
  return { query, indices };
}

function buildHttpRequest(values) {
  const http = values.http;
  return http;
}
