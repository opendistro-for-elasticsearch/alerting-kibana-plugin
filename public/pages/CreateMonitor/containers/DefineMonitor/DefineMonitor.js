/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { Component, Fragment } from 'react';
import _ from 'lodash';
import chrome from 'ui/chrome';
import PropTypes from 'prop-types';
import { EuiSpacer, EuiButton, EuiText, EuiCallOut } from '@elastic/eui';
import ContentPanel from '../../../../components/ContentPanel';
import VisualGraph from '../../components/VisualGraph';
import ExtractionQuery from '../../components/ExtractionQuery';
import MonitorExpressions from '../../components/MonitorExpressions';
import QueryPerformance from '../../components/QueryPerformance';
import MonitorDefinition from '../../components/MonitorDefinition';
import MonitorIndex from '../MonitorIndex';
import MonitorTimeField from '../../components/MonitorTimeField';
import { formikToMonitor } from '../CreateMonitor/utils/formikToMonitor';
import { getPathsPerDataType } from './utils/mappings';
import { buildSearchRequest } from './utils/searchRequests';
import { SEARCH_TYPE, ES_AD_PLUGIN } from '../../../../utils/constants';
import AnomalyDetectors from '../AnomalyDetectors/AnomalyDetectors';

function renderEmptyMessage(message) {
  return (
    <div style={{ padding: '20px', border: '1px solid #D9D9D9', borderRadius: '5px' }}>
      <div
        style={{ display: 'flex', alignItems: 'center', justifyContent: 'center', height: '450px' }}
      >
        <div>{message}</div>
      </div>
    </div>
  );
}

const propTypes = {
  values: PropTypes.object.isRequired,
  httpClient: PropTypes.func.isRequired,
  errors: PropTypes.object,
};
const defaultProps = {
  errors: {},
};

class DefineMonitor extends Component {
  constructor(props) {
    super(props);

    this.state = {
      dataTypes: {},
      performanceResponse: null,
      response: null,
      formikSnapshot: this.props.values,
      plugins: [],
    };

    this.renderGraph = this.renderGraph.bind(this);
    this.onRunQuery = this.onRunQuery.bind(this);
    this.resetResponse = this.resetResponse.bind(this);
    this.onQueryMappings = this.onQueryMappings.bind(this);
    this.queryMappings = this.queryMappings.bind(this);
    this.renderVisualMonitor = this.renderVisualMonitor.bind(this);
    this.renderExtractionQuery = this.renderExtractionQuery.bind(this);
    this.renderAnomalyDetector = this.renderAnomalyDetector.bind(this);
    this.getMonitorContent = this.getMonitorContent.bind(this);
    this.getPlugins = this.getPlugins.bind(this);
    this.showPluginWarning = this.showPluginWarning.bind(this);
    this.isDarkMode = chrome.getUiSettingsClient().get('theme:darkMode') || false;
  }

  componentDidMount() {
    this.getPlugins();
    const { searchType, index, timeField } = this.props.values;
    const isGraph = searchType === SEARCH_TYPE.GRAPH;
    const hasIndices = !!index.length;
    const hasTimeField = !!timeField;
    if (isGraph && hasIndices) {
      this.onQueryMappings();
      if (hasTimeField) this.onRunQuery();
    }
  }

  componentDidUpdate(prevProps) {
    const {
      searchType: prevSearchType,
      index: prevIndex,
      timeField: prevTimeField,
    } = prevProps.values;
    const { searchType, index, timeField } = this.props.values;
    const isGraph = searchType === SEARCH_TYPE.GRAPH;
    const hasIndices = !!index.length;
    // If customer is defining query through extraction query, then they are manually running their own queries
    // Below logic is for customers defining queries through graph/visual way.
    if (isGraph && hasIndices) {
      // If current query type is graph and there are indices selected, then we want to query new index mappings if
      // a) previous query type was query (to get the first load of mappings)
      // b) different indices, to get new mappings
      const wasQuery = prevSearchType === SEARCH_TYPE.QUERY;
      const diffIndices = prevIndex !== index;
      if (wasQuery || diffIndices) {
        this.onQueryMappings();
      }
      // If there is a timeField selected, then we want to run the query if
      // a) previous query type was query (to get first run executed)
      // b) different indices, to get new data
      // c) different time fields, to aggregate on new data/axis
      const diffTimeFields = prevTimeField !== timeField;
      const hasTimeField = !!timeField;
      if (hasTimeField) {
        if (wasQuery || diffIndices || diffTimeFields) this.onRunQuery();
      }
    }
  }

  async getPlugins() {
    const { httpClient } = this.props;
    try {
      const pluginsResponse = await httpClient.get('../api/alerting/_plugins');
      if (pluginsResponse.data.ok) {
        this.setState({ plugins: pluginsResponse.data.resp.map(plugin => plugin.component) });
      } else {
        console.error('There was a problem getting plugins list');
      }
    } catch (e) {
      console.error('There was a problem getting plugins list', e);
    }
  }

  renderGraph() {
    const { errors } = this.props;
    return (
      <Fragment>
        <EuiText size="xs">
          <strong>Create a monitor for</strong>
        </EuiText>
        <EuiSpacer size="s" />
        <MonitorExpressions
          onRunQuery={this.onRunQuery}
          dataTypes={this.state.dataTypes}
          ofEnabled={this.props.values.aggregationType !== 'count'}
        />
        <EuiSpacer size="s" />
        {errors.where ? (
          renderEmptyMessage('Invalid input in WHERE filter. Remove WHERE filter or adjust filter ')
        ) : (
          <VisualGraph
            values={this.state.formikSnapshot}
            fieldName={_.get(this.props.values, 'fieldName[0].label', 'Select a field')}
            response={this.state.response}
          />
        )}
      </Fragment>
    );
  }

  async onRunQuery() {
    const { httpClient, values } = this.props;
    const formikSnapshot = _.cloneDeep(values);

    // If we are running a visual graph query, then we need to run two separate queries
    // 1. The actual query that will be saved on the monitor, to get accurate query performance stats
    // 2. The UI generated query that gets [BUCKET_COUNT] times the aggregated buckets to show past history of query
    // If the query is an extraction query, we can use the same query for results and query performance
    const searchRequests = [buildSearchRequest(values)];
    if (values.searchType === SEARCH_TYPE.GRAPH) {
      searchRequests.push(buildSearchRequest(values, false));
    }

    try {
      const promises = searchRequests.map(searchRequest => {
        // Fill in monitor name in case it's empty (in create workflow)
        // Set triggers to empty array so they are not executed (if in edit workflow)
        // Set input search to query/graph query and then use execute API to fill in period_start/period_end
        const monitor = formikToMonitor(values);
        _.set(monitor, 'name', 'TEMP_MONITOR');
        _.set(monitor, 'triggers', []);
        _.set(monitor, 'inputs[0].search', searchRequest);
        return httpClient.post('../api/alerting/monitors/_execute', monitor);
      });

      const [queryResponse, optionalResponse] = await Promise.all(promises);

      if (queryResponse.data.ok) {
        const response = _.get(queryResponse.data.resp, 'input_results.results[0]');
        // If there is an optionalResponse use it's results, otherwise use the original response
        const performanceResponse = optionalResponse
          ? _.get(optionalResponse, 'data.resp.input_results.results[0]', null)
          : response;
        this.setState({ response, formikSnapshot, performanceResponse });
      } else {
        console.error('There was an error running the query', queryResponse.data.resp);
        this.setState({ response: null, formikSnapshot: null, performanceResponse: null });
      }
    } catch (err) {
      console.error('There was an error running the query', err);
    }
  }

  resetResponse() {
    this.setState({ response: null, performanceResponse: null });
  }

  async onQueryMappings() {
    const index = this.props.values.index.map(({ label }) => label);
    try {
      const mappings = await this.queryMappings(index);
      const dataTypes = getPathsPerDataType(mappings);
      this.setState({ dataTypes });
    } catch (err) {
      console.error('There was an error getting mappings for query', err);
    }
  }

  async queryMappings(index) {
    if (!index.length) {
      return {};
    }

    try {
      const response = await this.props.httpClient.post('../api/alerting/_mappings', { index });
      if (response.data.ok) {
        return response.data.resp;
      }
      return {};
    } catch (err) {
      throw err;
    }
  }
  renderVisualMonitor() {
    const { httpClient, values } = this.props;
    const { index, timeField } = values;
    const { dataTypes, performanceResponse } = this.state;
    let content = null;
    if (index.length) {
      content = timeField
        ? this.renderGraph()
        : renderEmptyMessage('You must specify a time field.');
    } else {
      content = renderEmptyMessage('You must specify an index.');
    }
    return {
      actions: [],
      content: (
        <React.Fragment>
          <MonitorIndex httpClient={httpClient} />
          <MonitorTimeField dataTypes={dataTypes} />
          <div style={{ padding: '0px 10px' }}>{content}</div>
          <EuiSpacer size="m" />
          <QueryPerformance response={performanceResponse} />
        </React.Fragment>
      ),
    };
  }
  renderExtractionQuery() {
    const { httpClient, values } = this.props;
    const { response, performanceResponse } = this.state;
    let invalidJSON = false;
    try {
      JSON.parse(values.query);
    } catch (e) {
      invalidJSON = true;
    }
    const runIsDisabled = invalidJSON || !values.index.length;
    let content = renderEmptyMessage('You must specify an index.');
    if (values.index.length) {
      content = (
        <ExtractionQuery
          response={JSON.stringify(response || '', null, 4)}
          isDarkMode={this.isDarkMode}
        />
      );
    }
    return {
      actions: [
        <EuiButton disabled={runIsDisabled} onClick={this.onRunQuery}>
          Run
        </EuiButton>,
      ],
      content: (
        <React.Fragment>
          <MonitorIndex httpClient={httpClient} />
          <div style={{ padding: '0px 10px' }}>{content}</div>
          <EuiSpacer size="m" />
          <QueryPerformance response={performanceResponse} />
        </React.Fragment>
      ),
    };
  }
  renderAnomalyDetector() {
    const { httpClient, values, detectorId } = this.props;
    return {
      actions: [],
      content: (
        <React.Fragment>
          <div style={{ padding: '0px 10px' }}>
            <AnomalyDetectors
              httpClient={httpClient}
              values={values}
              renderEmptyMessage={renderEmptyMessage}
              detectorId={detectorId}
            />
          </div>
        </React.Fragment>
      ),
    };
  }

  getMonitorContent() {
    const { values } = this.props;
    switch (values.searchType) {
      case SEARCH_TYPE.AD:
        return this.renderAnomalyDetector();
      case SEARCH_TYPE.GRAPH:
        return this.renderVisualMonitor();
      default:
        return this.renderExtractionQuery();
    }
  }
  showPluginWarning() {
    const { values } = this.props;
    const { plugins } = this.state;
    return values.searchType == SEARCH_TYPE.AD && plugins.indexOf(ES_AD_PLUGIN) == -1;
  }

  render() {
    const monitorContent = this.getMonitorContent();
    return (
      <ContentPanel
        title="Define monitor"
        titleSize="s"
        bodyStyles={{ padding: 'initial' }}
        actions={monitorContent.actions}
      >
        {this.showPluginWarning()
          ? [
              <EuiCallOut
                color="warning"
                title="Anomaly detector plugin is not installed on Elasticsearch, This monitor will not functional properly."
                iconType="help"
                size="s"
              />,
              <EuiSpacer size="s" />,
            ]
          : null}
        <MonitorDefinition resetResponse={this.resetResponse} plugins={this.state.plugins} />
        {monitorContent.content}
      </ContentPanel>
    );
  }
}

DefineMonitor.propTypes = propTypes;
DefineMonitor.defaultProps = defaultProps;

export default DefineMonitor;
