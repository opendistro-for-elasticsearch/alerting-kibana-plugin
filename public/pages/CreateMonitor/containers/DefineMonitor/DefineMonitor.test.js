/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { shallow } from 'enzyme';

import DefineMonitor from './DefineMonitor';
import { FORMIK_INITIAL_VALUES } from '../CreateMonitor/utils/constants';
import { httpClientMock } from '../../../../../test/mocks';

function getShallowWrapper(customProps = {}) {
  return shallow(
    <DefineMonitor values={FORMIK_INITIAL_VALUES} httpClient={httpClientMock} {...customProps} />
  );
}

describe('DefineMonitor', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    httpClientMock.post.mockResolvedValue({ data: { ok: true, resp: [] } });
    httpClientMock.get.mockResolvedValue({ data: { ok: true, resp: [] } });
  });

  test('renders', () => {
    const wrapper = getShallowWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  test('mounting should not call onQueryMappings or onRunQuery if not graph type', () => {
    const onQueryMappings = jest.spyOn(DefineMonitor.prototype, 'onQueryMappings');
    const onRunQuery = jest.spyOn(DefineMonitor.prototype, 'onRunQuery');
    const values = { ...FORMIK_INITIAL_VALUES, searchType: 'query', timeField: '@timestamp' };
    const wrapper = getShallowWrapper({ values });
    expect(onQueryMappings).not.toHaveBeenCalled();
    expect(onRunQuery).not.toHaveBeenCalled();
  });

  test('mounting should not call onQueryMappings or onRunQuery if no indices selected', () => {
    const onQueryMappings = jest.spyOn(DefineMonitor.prototype, 'onQueryMappings');
    const onRunQuery = jest.spyOn(DefineMonitor.prototype, 'onRunQuery');
    const values = {
      ...FORMIK_INITIAL_VALUES,
      searchType: 'graph',
      timeField: '@timestamp',
      index: [],
    };
    const wrapper = getShallowWrapper({ values });
    expect(onQueryMappings).not.toHaveBeenCalled();
    expect(onRunQuery).not.toHaveBeenCalled();
  });

  test('mounting should call getPlugins', () => {
    const getPlugins = jest.spyOn(DefineMonitor.prototype, 'getPlugins');
    const values = { ...FORMIK_INITIAL_VALUES, searchType: 'ad' };
    const wrapper = getShallowWrapper({ values });
    expect(getPlugins).toHaveBeenCalled();
  });
  test('should show warning in case of Ad monitor and plugin is not installed', () => {
    const getPlugins = jest.spyOn(DefineMonitor.prototype, 'getPlugins');
    const values = { ...FORMIK_INITIAL_VALUES, searchType: 'ad' };
    const wrapper = getShallowWrapper({ values });
    expect(wrapper).toMatchSnapshot();
  });

  test('mounting should only call onQueryMappings if graph type and indices are selected, but no timefield', () => {
    const onQueryMappings = jest.spyOn(DefineMonitor.prototype, 'onQueryMappings');
    const onRunQuery = jest.spyOn(DefineMonitor.prototype, 'onRunQuery');
    const values = {
      ...FORMIK_INITIAL_VALUES,
      searchType: 'graph',
      timeField: '',
      index: [{ label: 'some_index' }],
    };
    const wrapper = getShallowWrapper({ values });
    expect(onQueryMappings).toHaveBeenCalled();
    expect(onRunQuery).not.toHaveBeenCalled();
  });

  test('mounting should call onQueryMappings and onRunQuery if graph type, indices are selected, and timefield exists', () => {
    const onQueryMappings = jest.spyOn(DefineMonitor.prototype, 'onQueryMappings');
    const onRunQuery = jest.spyOn(DefineMonitor.prototype, 'onRunQuery');
    const values = {
      ...FORMIK_INITIAL_VALUES,
      searchType: 'graph',
      timeField: '@timestamp',
      index: [{ label: 'some_index' }],
    };
    const wrapper = getShallowWrapper({ values });
    expect(onQueryMappings).toHaveBeenCalled();
    expect(onRunQuery).toHaveBeenCalled();
  });

  test('should call onQueryMappings and onRunQuery when switching from query type to graph type', () => {
    const onQueryMappings = jest.spyOn(DefineMonitor.prototype, 'onQueryMappings');
    const onRunQuery = jest.spyOn(DefineMonitor.prototype, 'onRunQuery');
    const values = {
      ...FORMIK_INITIAL_VALUES,
      searchType: 'query',
      timeField: '@timestamp',
      index: [{ label: 'some_index' }],
    };
    const wrapper = getShallowWrapper({ values });
    expect(onQueryMappings).not.toHaveBeenCalled();
    expect(onRunQuery).not.toHaveBeenCalled();
    wrapper.setProps({ httpClient: httpClientMock, values: { ...values, searchType: 'graph' } });
    wrapper.update();
    expect(onQueryMappings).toHaveBeenCalled();
    expect(onRunQuery).toHaveBeenCalled();
  });

  test('should call onQueryMappings and onRunQuery when changing indices', () => {
    const onQueryMappings = jest.spyOn(DefineMonitor.prototype, 'onQueryMappings');
    const onRunQuery = jest.spyOn(DefineMonitor.prototype, 'onRunQuery');
    const values = {
      ...FORMIK_INITIAL_VALUES,
      searchType: 'graph',
      timeField: '@timestamp',
      index: [],
    };
    const wrapper = getShallowWrapper({ values });
    expect(onQueryMappings).not.toHaveBeenCalled();
    expect(onRunQuery).not.toHaveBeenCalled();
    wrapper.setProps({
      httpClient: httpClientMock,
      values: { ...values, index: [{ label: 'some_index' }] },
    });
    wrapper.update();
    expect(onQueryMappings).toHaveBeenCalled();
    expect(onRunQuery).toHaveBeenCalled();
  });

  test('should call onRunQuery when changing timefield', () => {
    const onRunQuery = jest.spyOn(DefineMonitor.prototype, 'onRunQuery');
    const values = {
      ...FORMIK_INITIAL_VALUES,
      searchType: 'graph',
      timeField: '@timestamp',
      index: [{ label: 'some_index' }],
    };
    const wrapper = getShallowWrapper({ values });
    expect(onRunQuery).toHaveBeenCalledTimes(1); // on mount
    wrapper.setProps({
      httpClient: httpClientMock,
      values: { ...values, timeField: 'different_time_field' },
    });
    wrapper.update();
    expect(onRunQuery).toHaveBeenCalledTimes(2); // on update
  });

  test('resetResponse should reset the query responses', () => {
    const wrapper = getShallowWrapper();
    wrapper.setState({ response: {}, performanceResponse: {} });
    expect(wrapper.state().response).not.toBe(null);
    expect(wrapper.state().performanceResponse).not.toBe(null);
    wrapper.instance().resetResponse();
    expect(wrapper.state().response).toBe(null);
    expect(wrapper.state().performanceResponse).toBe(null);
  });
});
