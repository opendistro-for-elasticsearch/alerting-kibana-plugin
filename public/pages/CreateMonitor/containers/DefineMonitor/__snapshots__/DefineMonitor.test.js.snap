// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefineMonitor renders 1`] = `
<ContentPanel
  actions={Array []}
  bodyStyles={
    Object {
      "padding": "initial",
    }
  }
  title="Define monitor"
  titleSize="s"
>
  <MonitorDefinition
    plugins={Array []}
    resetResponse={[Function]}
  />
  <MonitorIndex
    httpClient={[MockFunction]}
  />
  <MonitorTimeField
    dataTypes={Object {}}
  />
  <div
    style={
      Object {
        "padding": "0px 10px",
      }
    }
  >
    <div
      style={
        Object {
          "border": "1px solid #D9D9D9",
          "borderRadius": "5px",
          "padding": "20px",
        }
      }
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": "450px",
            "justifyContent": "center",
          }
        }
      >
        <div>
          You must specify an index.
        </div>
      </div>
    </div>
  </div>
  <EuiSpacer
    size="m"
  />
  <QueryPerformance
    response={null}
  />
</ContentPanel>
`;

exports[`DefineMonitor should show warning in case of Ad monitor and plugin is not installed 1`] = `
<ContentPanel
  actions={Array []}
  bodyStyles={
    Object {
      "padding": "initial",
    }
  }
  title="Define monitor"
  titleSize="s"
>
  <EuiCallOut
    color="warning"
    iconType="help"
    size="s"
    title="Anomaly detector plugin is not installed on Elasticsearch, This monitor will not functional properly."
  />
  <EuiSpacer
    size="s"
  />
  <MonitorDefinition
    plugins={Array []}
    resetResponse={[Function]}
  />
  <div
    style={
      Object {
        "padding": "0px 10px",
      }
    }
  >
    <AnomalyDetectors
      httpClient={[MockFunction]}
      renderEmptyMessage={[Function]}
      values={
        Object {
          "aggregationType": "count",
          "bucketUnitOfTime": "h",
          "bucketValue": 1,
          "cronExpression": "0 */1 * * *",
          "daily": 0,
          "detectorId": "",
          "disabled": false,
          "fieldName": Array [],
          "frequency": "interval",
          "groupedOverFieldName": "bytes",
          "groupedOverTop": 5,
          "index": Array [],
          "monthly": Object {
            "day": 1,
            "type": "day",
          },
          "name": "",
          "overDocuments": "all documents",
          "period": Object {
            "interval": 1,
            "unit": "MINUTES",
          },
          "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
          "searchType": "ad",
          "timeField": "",
          "timezone": Array [],
          "weekly": Object {
            "fri": false,
            "mon": false,
            "sat": false,
            "sun": false,
            "thur": false,
            "tue": false,
            "wed": false,
          },
          "where": Object {
            "fieldName": Array [],
            "fieldRangeEnd": 0,
            "fieldRangeStart": 0,
            "fieldValue": "",
            "operator": "is",
          },
        }
      }
    />
  </div>
</ContentPanel>
`;
